packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx

```diff
@@ -0,0 +1,190 @@
+ * limitations under the License.
+ */
+
+import {
+  Dialog,
+  PanelDropZone,
+  PanelFormBooleanField,
+  PanelForm,
+  PanelDivider,
+  Modal,
+  ModalBody,
+  ModalFooter,
+  ModalHeader,
+  PanelFormListItems,
+  PanelFormSection,
+} from '@finos/legend-art';
+import {
+  PRIMITIVE_TYPE,
+  TYPICAL_MULTIPLICITY_TYPE,
+  type ValueSpecification,
+} from '@finos/legend-graph';
+import { guaranteeNonNullable } from '@finos/legend-shared';
+import { observer } from 'mobx-react-lite';
+import { useCallback } from 'react';
+import { useDrop } from 'react-dnd';
+import {
+  type QueryBuilderParameterDragSource,
+  QUERY_BUILDER_PARAMETER_DND_TYPE,
+} from '../../stores/QueryBuilderParametersState.js';
+import type { QueryBuilderState } from '../../stores/QueryBuilderState.js';
+import type { QueryBuilderWatermarkState } from '../../stores/watermark/QueryBuilderWatermarkState.js';
+import { VariableExpressionViewer } from '../QueryBuilderParametersPanel.js';
+import { BasicValueSpecificationEditor } from '../shared/BasicValueSpecificationEditor.js';
+
+const WatermarkValueEditor = observer(
+  (props: {
+    watermarkValue: ValueSpecification;
+    watermarkState: QueryBuilderWatermarkState;
+  }) => {
+    const { watermarkValue, watermarkState } = props;
+
+    const graph = watermarkState.queryBuilderState.graphManagerState.graph;
+
+    const handleDrop = useCallback(
+      (item: QueryBuilderParameterDragSource): void => {
+        watermarkState.setValue(item.variable.parameter);
+      },
+      [watermarkState],
+    );
+
+    const [{ isParameterValueDragOver }, dropTargetConnector] = useDrop<
+      QueryBuilderParameterDragSource,
+      void,
+      { isParameterValueDragOver: boolean }
+    >(
+      () => ({
+        accept: [QUERY_BUILDER_PARAMETER_DND_TYPE],
+        drop: (item, monitor): void => {
+          const itemType = item.variable.parameter.genericType?.value.rawType;
+          if (
+            !monitor.didDrop() &&
+            // Only allows parameters with muliplicity 1 and type string
+            item.variable.parameter.multiplicity ===
+              graph.getTypicalMultiplicity(TYPICAL_MULTIPLICITY_TYPE.ONE) &&
+            item.variable.parameter.genericType?.value.rawType.name ===
+              'String' &&
+            itemType
+          ) {
+            handleDrop(item);
+          } // prevent drop event propagation to accomondate for nested DnD
+        },
+        collect: (monitor) => ({
+          isParameterValueDragOver: monitor.isOver({
+            shallow: true,
+          }),
+        }),
+      }),
+      [handleDrop],
+    );
+
+    return (
+      <PanelFormSection>
+        <div className="query-builder__parameter-editor">
+          <PanelDropZone
+            isDragOver={isParameterValueDragOver}
+            dropTargetConnector={dropTargetConnector}
+          >
+            <BasicValueSpecificationEditor
+              valueSpecification={guaranteeNonNullable(watermarkValue)}
```

Comment:
you shouldnt need the guarantNonNullable if the input for this component is `watermarkValue: ValueSpecification`

Reply:
oh makes sense!